service/src/test/java/bio/terra/workspace/service/workspace/WorkspaceServiceTest.java

```diff
@@ -277,46 +280,53 @@ void createAndDeleteWorkspace() {
+    retrySteps.put(CreateWorkspaceStep.class.getName(), StepStatus.STEP_RESULT_FAILURE_RETRY);
+    FlightDebugInfo debugInfo = FlightDebugInfo.newBuilder().doStepFailures(retrySteps).build();
     jobService.setFlightDebugInfoForTest(debugInfo);
 
-    workspaceService.deleteWorkspace(request.workspaceId(), USER_REQUEST);
-    assertThrows(
-        WorkspaceNotFoundException.class,
-        () -> workspaceService.getWorkspace(request.workspaceId(), USER_REQUEST));
+    UUID createdId = workspaceService.createWorkspace(request, USER_REQUEST);
+    assertEquals(createdId, request.workspaceId());
   }
 
   @Test
-  void deleteWorkspaceUndoFailure() {
+  void createRawlsWorkspaceDoSteps() throws InterruptedException {
     WorkspaceRequest request =
         defaultRequestBuilder(UUID.randomUUID())
-            .workspaceStage(WorkspaceStage.MC_WORKSPACE)
+            .workspaceStage(WorkspaceStage.RAWLS_WORKSPACE)
             .build();
-    workspaceService.createWorkspace(request, USER_REQUEST);
+    // Ensure the auth check in CheckSamWorkspaceAuthzStep always succeeds.
+    doReturn(true).when(mockSamService).isAuthorized(any(), any(), any(), any());
+    Map<String, StepStatus> retrySteps = new HashMap<>();
+    retrySteps.put(
+        CheckSamWorkspaceAuthzStep.class.getName(), StepStatus.STEP_RESULT_FAILURE_RETRY);
+    retrySteps.put(CreateWorkspaceStep.class.getName(), StepStatus.STEP_RESULT_FAILURE_RETRY);
+    FlightDebugInfo debugInfo = FlightDebugInfo.newBuilder().doStepFailures(retrySteps).build();
+    jobService.setFlightDebugInfoForTest(debugInfo);
 
-    // Test "undo" flight with lastStepFailure. Because deletion cannot be undone, this workspace
-    // will still be deleted after "undoing" the flight.
-    FlightDebugInfo debugInfo = FlightDebugInfo.newBuilder().lastStepFailure(true).build();
+    UUID createdId = workspaceService.createWorkspace(request, USER_REQUEST);
+    assertEquals(createdId, request.workspaceId());
+  }
+
+  @Test
+  void createWorkspaceUndoSteps() {
+    WorkspaceRequest request = defaultRequestBuilder(UUID.randomUUID()).build();
+    // Retry undo steps once and fail at the end of the flight.
+    Map<String, StepStatus> retrySteps = new HashMap<>();
+    retrySteps.put(CreateWorkspaceAuthzStep.class.getName(), StepStatus.STEP_RESULT_FAILURE_RETRY);
+    retrySteps.put(CreateWorkspaceStep.class.getName(), StepStatus.STEP_RESULT_FAILURE_RETRY);
+    FlightDebugInfo debugInfo =
+        FlightDebugInfo.newBuilder().undoStepFailures(retrySteps).lastStepFailure(true).build();
     jobService.setFlightDebugInfoForTest(debugInfo);
 
-    // When a flight fails with no error message (e.g. because of debugInfo), Stairway will return
-    // an InvalidResultStateException.
+    // JobService throws a InvalidResultStateException when a synchronous flight fails without an
+    // exception, which occurs when a flight fails via debugInfo.
```

Comment:
nit: Not quite following the "synchronous flight" part of this comment -- I usually think of endpoints as a/sync, but not the flights themselves.

Reply:
Yep, this was really shorthand - flights aren't really sync or async, but since each flight is generally only called from one (sync or async) service endpoint, I tend to think of them as sync or async flights. 
Clarified this comment everywhere